### ISSUE TITLE
Does HMR only replace module

### ISSUE THREAD

**Issue by eyalhakim** [enten/angular-universal#34]
When changing a component under a module, does only this module reloads? I see in the console that HMR logs the entire chain all the way to main.ts.

**Comment by enten** [comment]
@eyalhakim 
Your question is more related to [webpack hmr](https://webpack.js.org/guides/hot-module-replacement/) than angular development.

I will try to resume whats happens in file under webpack hmr mode:
* webpack computes delta between last build and file change and create a piece of code called "hot chunk" ;
* the hot chunk can be apply to the last build to have the same result as a full rebuild (but without rebuilding everything) ;
* but webpack doesn't know about the change and doesn't know how to apply the change (hot chunk) ;
* it's the developer responsibility to tell webpack how apply hot chunk by using `module.hot` instructions ;
* so webpack checks for `module.hot` instructions in file which chnage ;
* if webpack found `module.hot` instructions: he call thems ;
* else webpack try to do same checks in parent until it hit the main entry ;
* if main entry doesn't have hot instructions too: webpack consider that it can apply hot chunk and log an hmr error.

_This repo support HMR or browser sides. In example below we will talks about hmr on browser side_.

So, if a change happen in `src/app/welcome/welcome.component.ts` for example: webpack found hot module instructions in browser entry point `main.ts` and apply the hot chunk.

For this reason you can see log below:
```
[HMR] Updated modules:
[HMR]  - ./src/app/welcome/welcome.component.ts
[HMR]  - ./src/app/welcome/welcome-routing.module.ts
[HMR]  - ./src/app/welcome/welcome.module.ts
[HMR]  - ./src/app/app-routing.module.ts
[HMR]  - ./src/app/app.module.ts
[HMR]  - ./src/app/app.browser.module.ts
[HMR]  - ./src/main.ts
[HMR] App is up to date.
```

But if you tells webpack how hot reload this file `./src/app/welcome/welcome.component.ts` but adding some `module.hot` instructions inside: webpack will apply chunk on WelcomeComponent only. And the log will be;

```
[HMR] Updated modules:
[HMR]  - ./src/app/welcome/welcome.component.ts
[HMR] App is up to date.
```

Now if you remove hot instructions from browser entry point and update a file (which has no hot modules instructions too): you will see warnings like below (in browser side but please that will throw an error in server side):

```
Ignored an update to unaccepted module ./src/app/welcome/welcome.component.ts
-> ./src/app/welcome/welcome-routing.module.ts
-> ./src/app/welcome/welcome.module.ts
-> ./src/app/app-routing.module.ts
-> ./src/app/app.module.ts
-> ./src/app/app.browser.module.ts
-> ./src/main.ts
-> 0
[HMR] The following modules couldn't be hot updated: (Full reload needed)
This is usually because the modules which have changed (and their parents)
do not know how to hot reload themselves.
See https://webpack.js.org/concepts/hot-module-replacement/ for more details.
[HMR]  - ./src/app/welcome/welcome.component.ts
```

**In conclusion**: if you don't write specific hot module instructions, on browser side, when a change happens, it will be the browser main.ts entry point which will apply hot chunk (and webpack will print the entire chain).

**Comment by eyalhakim** [comment]
Got it thank you

**Comment by chaitanyagurrapu** [comment]
> if webpack found module.hot instructions: he call thems 

@enten Can you please provide an example of how to provide instructions for module.hot to only reload one component?

**Comment by enten** [comment]
@chaitanyagurrapu  
Reloading only one component can be hard if this component is in module which expose lot of things.

A fast way to achieve that is to create a module for your component.

In that way you easily manage hmr of your component module as shown in angular hmr example:
https://github.com/PatrickJS/angular-hmr#angular-hot-module-replacement
